### PR TITLE
fuzz: have package_rbf always make small txns

### LINKED ISSUE #30241: fuzz: timeout/oom in package_rbf
I stumbled upon a fuzz seed that oom'ed in CI for `package_rbf`. Since the seed appears to be too big to paste, I provide the output of `$ base64 95a3b1b42d43ac7190f24ff116b2cd6515e7a566 > /tmp/base64-95a3b1b4-oom.txt`:
[base64-95a3b1b4-oom.txt](https://github.com/user-attachments/files/15685350/base64-95a3b1b4-oom.txt)

If the prior format is unsatisfactory, the problematic seed `95a3b1b42d43ac7190f24ff116b2cd6515e7a566` is also in this commit to qa-assets: https://github.com/bitcoin-core/qa-assets/pull/186/commits/8d624b44419f46bcb5117baa16047dbf4bdb339f

_Originally posted by @murchandamus in https://github.com/bitcoin/bitcoin/issues/28812#issuecomment-2153274374_

### PR DESCRIPTION
**Pull request by instagibbs**
hopefully resolves https://github.com/bitcoin/bitcoin/issues/30241

The fuzz target is generating a large amount of
transactions, but the core of the logic is
ConsumeTxMemPoolEntry making the mempool
entries for adding to the mempool. Since
ConsumeTxMemPoolEntry generates its own transaction "vsize",
we can improve efficiency of the target
by explicitly creating very small transactions,
reducing the hashing and memory burden.

### CODE REVIEW

**Comment by instagibbs** [comment]
> Oh, I guess indirectly via sig_op_cost?

Right, sorry! 

https://github.com/bitcoin/bitcoin/pull/29879 previously touched on this

File: src/test/fuzz/rbf.cpp [hunk 1]
```diff
@@ -113,15 +115,13 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
...
-        std::optional<CMutableTransaction> parent = ConsumeDeserializable<CMutableTransaction>(fuzzed_data_provider, TX_WITH_WITNESS);
-        if (!parent) {
-            return;
-        }
+        CMutableTransaction parent;
         assert(iter <= g_outpoints.size());
-        parent->vin.resize(1);
-        parent->vin[0].prevout = g_outpoints[iter++];
+        parent.vin.resize(1);
+        parent.vin[0].prevout = g_outpoints[iter++];
```
Comment (hodlinator) [inline src/test/fuzz/rbf.cpp:121]: Might as well
```suggestion
        parent.vin.emplace_back(g_outpoints[iter++]);
```

File: src/test/fuzz/rbf.cpp [hunk 2]
```diff
@@ -113,15 +115,13 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
     LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), NUM_ITERS)
     {
         // Make sure txns only have one input, and that a unique input is given to avoid circular references
```
Comment (hodlinator) [inline src/test/fuzz/rbf.cpp:117]: I understand this comment as explaining why a random number of `vin`s of `parent` generated from fuzz data are overwritten with one input. Time to change/remove?

File: src/test/fuzz/rbf.cpp [hunk 3]
```diff
@@ -130,10 +130,10 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
             break;
         }
         pool.addUnchecked(parent_entry);
-        if (fuzzed_data_provider.ConsumeBool() && !child->vin.empty()) {
-            child->vin[0].prevout = COutPoint{mempool_txs.back().GetHash(), 0};
+        if (fuzzed_data_provider.ConsumeBool()) {
+            child.vin[0].prevout = COutPoint{mempool_txs.back().GetHash(), 0};
```
Comment (hodlinator) [inline src/test/fuzz/rbf.cpp:134]: Good that you removed the unnecessary `vin.empty()` check.

Until the first time `ConsumeBool()` returns true here, we'll be using a `child` with a default-initialized `vin[0]`. This was true before the change as well. Does some later code hook up default-constructed `CTxIn` to coinbase transactions or something?

It seems unnecessary to have created a `parent` at all if the `ConsumeBool()` returns false. But I guess it's good to test adding unrelated transactions between each `child`.

**Review by hodlinator — COMMENTED** [review]
ACK 4ccb3d6d0d576d32da8a1b9c6e70962cbd0f19fe

**Review by glozow — COMMENTED** [review]
ACK 4ccb3d6d0d576d32da8a1b9c6e70962cbd0f19fe

Agree value of `GetTxSize()` is chosen by `ConsumeTxMemPoolEntry` and conflicts aren't calculated using outpoints, so there's no need to make mempool transactions through `ConsumeDeserializable<CMutableTransaction>`.

seed 95a3b1b42d43ac7190f24ff116b2cd6515e7a566 didn't give me an oom but took 1981ms / now takes 71ms.

**Comment by murchandamus** [comment]
This change in the fuzzer seems to have reduced the coverage of the `package_rbf` fuzzer with the current `qa-assets` fuzz inputs by about 600 LOC. I’ll make a PR to update the fuzz inputs.

**Pull request merged**